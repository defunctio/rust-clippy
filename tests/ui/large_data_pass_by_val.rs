#![warn(clippy::large_data_pass_by_val)]

#[derive(Copy, Clone)]
struct A {
    x: [u64; 0x200],
}

struct B {
    x: [u64; 0x200],
}

trait C {
    fn good(&self, arg: &A);
    fn bad(&self, arg: A);
}

impl C for B {
    fn good(&self, arg: &A) {
        unimplemented!()
    }
    fn bad(&self, arg: A) {
        unimplemented!()
    }
}

impl B {
    fn do_bad(&self, arg: A) {
        unimplemented!()
    }
    fn do_good(&self, arg: &A) {
        unimplemented!()
    }
}

fn pass_by_val(arg: A) {
    let sum = arg.x.iter().sum::<u64>();
    println!("{}", sum);
}

fn pass_by_ref(arg: &A) {
    let sum = arg.x.iter().sum::<u64>();
    println!("{}", sum);
}

fn main() {
    let a = A { x: [0; 0x200] };
    pass_by_val(a);
    pass_by_ref(&a);
    let b = B { x: [0; 0x200] };
    b.do_bad(a);
    b.good(&a);
    b.bad(a);
}
