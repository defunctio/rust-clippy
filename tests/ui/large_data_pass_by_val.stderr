error: this argument (4096 byte) is passed by value, but would be more efficient if passed by ref (limit: 16 byte)
  --> $DIR/large_data_pass_by_val.rs:14:24
   |
LL |     fn bad(&self, arg: A);
   |                        ^ help: consider passing by ref instead: `A`
   |
   = note: `-D clippy::large-data-pass-by-val` implied by `-D warnings`

error: this argument (4096 byte) is passed by value, but would be more efficient if passed by ref (limit: 16 byte)
  --> $DIR/large_data_pass_by_val.rs:27:27
   |
LL |     fn do_bad(&self, arg: A) {
   |                           ^ help: consider passing by ref instead: `A`

error: this argument (4096 byte) is passed by value, but would be more efficient if passed by ref (limit: 16 byte)
  --> $DIR/large_data_pass_by_val.rs:35:21
   |
LL | fn pass_by_val(arg: A) {
   |                     ^ help: consider passing by ref instead: `A`

error: aborting due to 3 previous errors

